tsec-libsodium/src/main/scala/tsec/libsodium/kdf/KeyDerivation.scala

```diff
@@ -0,0 +1,37 @@
+package tsec.libsodium.kdf
+
+import cats.effect.Sync
+import tsec.common._
+import tsec.libsodium.ScalaSodium
+import tsec.libsodium.cipher.SodiumKey
+import tsec.libsodium.pk.PrivateKey
+
+sealed trait KeyDerivation
+
+object KeyDerivation {
+
+  def generateKey[F[_]](implicit F: Sync[F], S: ScalaSodium): F[PrivateKey[KeyDerivation]] = F.delay {
+    val masterKey = PrivateKey[KeyDerivation](new Array[Byte](ScalaSodium.crypto_kdf_KEYBYTES))
+    S.crypto_kdf_keygen(masterKey)
+    masterKey
+  }
+
+  def deriveKey[F[_]](
+      masterKey: PrivateKey[KeyDerivation],
+      keyLength: Int,
+      id: Int,
+      context: String
+  )(implicit F: Sync[F], S: ScalaSodium): F[SodiumKey[KeyDerivation]] = F.delay {
+    if (ScalaSodium.crypto_kdf_BYTES_MIN > keyLength || keyLength > ScalaSodium.crypto_kdf_BYTES_MAX)
+      throw KeyLengthError(keyLength)
+
+    val ctx = context.utf8Bytes
+    if (ctx.length != ScalaSodium.crypto_kdf_CONTEXTBYTES)
+      throw ContextBytesError(ctx.length)
```

Comment:
I know it's caught by the `delay` call, but I still dislike using `throw` in general. `Sync` (`ApplicativeError`) has `raiseError` for things like this, IIRC.

Reply:
You are right, but the signature is `raiseError[A](e: Throwable): F[A]` and we wrapped it inside `F.delay`, I think there is no way arround ?

Reply:
You can do like:

```scala
def checkKeyLength[F[_]: ApplicativeError[?, Throwable]](kl: Int): F[Unit] =
  if (SS.crypto_kdf_BYTES_MIN > kl || kl > SS.crypto_kdf_BYTES_MAX)
    ApplicativeError[F].raiseError(kl)
  else ().pure[F]

def checkContextBytes = ...
// then...
val ctx = context.utf8Bytes
checkKeyLength(keyLength) *> checkContextBytes(ctx) *> F.delay {
  ...
}
```

Reply:
I don't like the 3 flatmaps personally. Maybe im just a stickler for performance, but that's at least 3 calls into an `IO` data structure + flatmaps.

I am personally guilty of `throw` in tsec-libsodium. That said, it's because some of the conditionals are _after_ a side effect.

Reply:
Lol I deleted the wrong comment accidentally.

I was gonna say, even if this style is less functional, it's most certainly more efficient (no flatmaps, less allocations), so I'm ok with it.